python-package/packager/pep517.py

```diff
@@ -0,0 +1,191 @@
+"""
+Custom build backend for XGBoost Python package.
+Builds source distribution and binary wheels, following PEP 517 / PEP 660.
+Re-uses components of Hatchling (https://github.com/pypa/hatch/tree/master/backend) for the sake
+of brevity.
+"""
+import dataclasses
+import logging
+import os
+import pathlib
+import sysconfig
+import tempfile
+from contextlib import contextmanager
+from typing import Any, Dict, Iterator, List, Optional, Union
+
+import hatchling.build
+
+from .build_config import BuildConfiguration
+from .nativelib import locate_local_libxgboost, locate_or_build_libxgboost
+from .sdist import copy_cpp_src_tree
+from .util import copy_with_logging, copytree_with_logging
+
+
+@contextmanager
+def cd(path: Union[str, pathlib.Path]) -> Iterator[str]:  # pylint: disable=C0103
+    """Temporarily change working directory"""
+    if isinstance(path, pathlib.Path):
+        path = str(path)
+    path = os.path.realpath(path)
+    cwd = os.getcwd()
+    os.chdir(path)
+    try:
+        yield path
+    finally:
+        os.chdir(cwd)
+
+
+def get_tag() -> str:
+    """Get appropate wheel tag, according to system"""
+    tag_platform = sysconfig.get_platform().replace("-", "_").replace(".", "_")
+    return f"py3-none-{tag_platform}"
+
+
+TOPLEVEL_DIR = pathlib.Path(__file__).parent.parent.absolute().resolve()
+logging.basicConfig(level=logging.INFO)
+
+
+def get_requires_for_build_wheel(
+    config_settings: Optional[Dict[str, Any]] = None
+) -> List[str]:
+    """A PEP 517 method. Delegate to Hatchling"""
+    return hatchling.build.get_requires_for_build_wheel(config_settings)
+
+
+def get_requires_for_build_sdist(
+    config_settings: Optional[Dict[str, Any]] = None
+) -> List[str]:
+    """A PEP 517 method. Delegate to Hatchling"""
+    return hatchling.build.get_requires_for_build_sdist(config_settings)
+
+
+def get_requires_for_build_editable(
+    config_settings: Optional[Dict[str, Any]] = None
+) -> List[str]:
+    """A PEP 517 method. Delegate to Hatchling"""
+    return hatchling.build.get_requires_for_build_editable(config_settings)
+
+
+def write_hatch_config(dest_dir: pathlib.Path, *, logger: logging.Logger) -> None:
+    """Write a small custom hook for Hatch, to set a custom tag."""
+    template = (
+        "from hatchling.builders.hooks.plugin.interface import BuildHookInterface\n"
+        "class CustomBuildHook(BuildHookInterface):\n"
+        "    def initialize(self, version, build_data):\n"
+        "        build_data['tag'] = '{tag}'\n"
+    )
```

Comment:
You should be able to write this a little more neatly using textwrap.dedent and a multiline string to get the desired formatting
```suggestion
    template = textwrap.dedent("""
        from hatchling.builders.hooks.plugin.interface import BuildHookInterface
        class CustomBuildHook(BuildHookInterface):
            def initialize(self, version, build_data):
                build_data['tag'] = '{tag}'"""
    )
```